main.go

```diff
@@ -12,6 +24,99 @@ func init() {
 	_ = flag.Set("logtostderr", "true")
 }
 
+func GetCurrentNode(clientset kubernetes.Clientset) (*corev1.Node, error) {
+	nodeName := os.Getenv("NODE_NAME")
+	if nodeName == "" {
+		return nil, errors.New("Environment variable NODE_NAME is not set")
+	}
+
+	node, err := clientset.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node: %w", err)
+	}
+
+	return node, nil
+}
+
+func UpdateNodeLabels(node *corev1.Node, instanceData metadata.InstanceData) {
+	klog.Infof("Updating node labels with Linode instance data: %v", instanceData)
+
+	node.Labels["linode_label"] = instanceData.Label
+	node.Labels["linode_id"] = strconv.Itoa(instanceData.ID)
+	node.Labels["linode_region"] = instanceData.Region
+	node.Labels["linode_type"] = instanceData.Type
+	node.Labels["linode_host"] = instanceData.HostUUID
+}
+
+func GetClientset() (*kubernetes.Clientset, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientset, nil
+}
+
+func StartDecorator(client metadata.Client, node *corev1.Node, interval time.Duration) {
+	instanceData, err := client.GetInstance(context.Background())
+	if err != nil {
+		klog.Errorf("Failed to get the initial instance data: %s", err.Error())
+	}
+
+	if instanceData != nil {
+		UpdateNodeLabels(node, *instanceData)
+	}
+
+	instanceWatcher := client.NewInstanceWatcher(
+		metadata.WatcherWithInterval(interval),
+	)
+
+	go instanceWatcher.Start(context.Background())
+
+	for {
+		select {
+		case data := <-instanceWatcher.Updates:
+			if instanceData != nil {
```

Comment:
Why do we care if `instanceData` is nil here? 

If we get nil instanceData to begin with we will never update the node labels, which doesn't seem desirable?

Reply:
I think it will be safer to check `nil` of a pointer. I think I can add a warning log into the else branch when the `nil` appears. Do you think that's a good idea?

Reply:
As in check that `data` is not nil? I can definitely see the benefit in that, adding an else branch that'll log if it is nil makes sense to me. I guess it's running customer side but even so, if they let support know that there's a lot of `nil` messages being sent on the channel that'd be something for us to look into!